### PR TITLE
11_3_real_promises

### PR DESCRIPTION
**Pull request by TheIronHobo**
### Real Promises
- Implementation of 11_1 exercise that uses real promises in place of async/await
- Weekly logs folder/files not included in commit

### CODE REVIEW

File: eloquentjs/11_3_real_promises/real_promises.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Returns promise of hourly security camera activity recorded for given day of week (0-6)
+ * @param {*} day 
+ * @returns 
+ */
+function activityTable(day) {
+    return new Promise((resolve, reject) => {
+        textFile('camera_logs.txt')
+        .then(file => Promise.resolve(fileToArray(file)))
+        .then(array => Promise.all(array.map(
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/real_promises.js]: is there a clearer name you could use than array?

File: eloquentjs/11_3_real_promises/real_promises.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Returns promise of hourly security camera activity recorded for given day of week (0-6)
+ * @param {*} day 
+ * @returns 
+ */
+function activityTable(day) {
+    return new Promise((resolve, reject) => {
+        textFile('camera_logs.txt')
+        .then(file => Promise.resolve(fileToArray(file)))
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/real_promises.js]: Is there a clearer name you could use than file? Is the thing it's storing really a file?

Also, I don't think you should be using Promise.resolve like this. I understand the intention, but you can just combine this then with the next then into a longer (but still short) little function.

File: eloquentjs/11_3_real_promises/src/gaussian.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+function gaussian(x, variance, center, scale = 1) {
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/src/gaussian.js]: I'd just include this in the generateLogs file. Lots of files means lots of context shifts.

File: eloquentjs/11_3_real_promises/src/fileToArray.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const fileToArray = file => file.split('\n').filter(j => j.length > 0);
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/src/fileToArray.js]: I'd probably combine this and textFile into a module named something like logFileHelpers

**Review by CalHinshaw — CHANGES_REQUESTED** [review]
Great start, promises are really hard to wrap your head around.

File: eloquentjs/11_3_real_promises/real_promises.js (hunk 2 above)
Comment (TheIronHobo) [reply eloquentjs/11_3_real_promises/real_promises.js]: Along with the refactor, I've taken a stab at renaming many of the variables in this section. It feels like a hard balance between being too descriptive and not specific enough.

File: eloquentjs/11_2_building_promise_all.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+function Promise_all(promises) {
```
Comment (CalHinshaw) [inline eloquentjs/11_2_building_promise_all.js:1]: I don't think this was supposed to end up here

File: eloquentjs/11_3_real_promises/real_promises.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Returns promise of hourly security camera activity recorded for given day of week (0-6)
+ * @param {*} day 
+ * @returns 
+ */
+function activityTable(day) {
+    return new Promise((resolve, reject) => {
+        textFile('camera_logs.txt')
+        .then(logManifest => Promise.all(listStringToArray(logManifest).map(logFileName => {
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/real_promises.js]: When you nest this many expressions it gets very hard to follow. Assignment can be used to make what's happening clearer. For example:
```js
.then(logManifest => {
  const weeklyLogPromises = listStringToArray(logManifest)
    .map(logFileName => textFile(`weekly_logs/${logFileName}`));
  return Promise.all(weeklyLogPromises)
})
```

File: eloquentjs/11_3_real_promises/real_promises.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            return textFile(`weekly_logs/${logFileName}`);
+        })))
+        .then(timeStampLists => {
+            logs = timeStampLists.map(timeStampList_week => {
```
Comment (CalHinshaw) [inline eloquentjs/11_3_real_promises/real_promises.js]: don't forget to use const here. js will often let you do things that behave strangely, and this can be one of those things in some runtimes.

**Review by CalHinshaw — APPROVED** [review]
Looks really good, good job!

File: eloquentjs/11_3_real_promises/real_promises.js (hunk 2 above)
Comment (CalHinshaw) [reply eloquentjs/11_3_real_promises/real_promises.js]: I don't think being too descriptive is a significant problem. What are you worried about?

Comment (TheIronHobo) [reply eloquentjs/11_3_real_promises/real_promises.js]: It's not a significant issue for me, but I sometimes worry that being overly-descriptive would make my code cumbersome to read and write. I had briefly considered naming the `logManifest` `cameraLogRawContents` but decided against it.

Comment (TheIronHobo) [reply eloquentjs/11_3_real_promises/real_promises.js]: I suppose `logManifest` feels like a better description of what I intend to use that variable _for_ whereas `cameraLogRawContents` feels like a better description of what it _is_.

Comment (CalHinshaw) [reply eloquentjs/11_3_real_promises/real_promises.js]: "Overly-descriptive" is somewhat loaded language that I think might be coloring your thinking. "more descriptive" is a truer reflection of reality. Being more descriptive in your code makes simpler code more cumbersome to write, more complex code easier to write, and all code much easier to read. Code is generally read much more than written, so that's the problem you should optimize for.

I think a good way to think about what makes good code is that it reads like a story. Imagine you're going to explain your code to someone. You want concepts to appear in the order you'd want to cover them in, and you'd want ideas (eg variables and functions) to have the same names you'd want to refer to them as. Learning to find the words and see those patterns is quite a bit of work, but is a worthwhile challenge - it will set you apart as an engineer.

When it comes to the name you had as `logManifest`, I think using the word manifest obscures what it is too much, and I don't think cameraLogRawContents is actually what it is. I would have called it `rawCameraLogFilenames` since that's what it is.
